### PR TITLE
#15: Create worker Docker container deployer and executor 

### PR DESCRIPTION
**Pull request by jakejack13**
### Summary

This PR creates the Docker container deployment system in the leader for creating and executing worker nodes in Docker containers on a local Docker daemon. 

Issue: #15 

Currently to run this PR for proper results:

1. `docker compose up` the `docker-compose.yml` file in the kafka directory 
2. Run the `leaders main` configuration

### CODE REVIEW

File: leaders/src/main/java/edu/cornell/CloseableSet.java
Comment (jakejack13) [inline leaders/src/main/java/edu/cornell/CloseableSet.java:7]: Missing documentation explaining class

File: leaders/src/main/java/edu/cornell/Main.java
Comment (jakejack13) [inline leaders/src/main/java/edu/cornell/Main.java:60]: Should fix now

File: leaders/src/main/java/edu/cornell/Main.java
Comment (jakejack13) [inline leaders/src/main/java/edu/cornell/Main.java:62]: For some reason, this doesn't seem to be running. Will need to look into this

File: leaders/src/main/java/edu/cornell/Main.java
Comment (jakejack13) [inline leaders/src/main/java/edu/cornell/Main.java:69]: This implementation consists of us creating an empty set of workers, then sticking it into the `WorkerRunner`, then adding workers to the worker set, then running the thing. This pattern absolutely sucks since it breaks the abstraction barrier. I'm going to change it for my own sanity

File: leaders/src/main/java/edu/cornell/worker/DockerWorker.java
Comment (jakejack13) [inline leaders/src/main/java/edu/cornell/worker/DockerWorker.java:141]: Add `done()` call here

**Review by jakejack13 — COMMENTED** [review]
Some comments for myself

File: leaders/src/main/java/edu/cornell/worker/DockerWorker.java
Comment (jakejack13) [inline leaders/src/main/java/edu/cornell/worker/DockerWorker.java:81]: Worker seems to run just fine on the host computer but as soon as it's in a Docker container, it can't connect to Kafka. I tried doing this port forwarding but doesn't seem to work. Will need to look into it more. 

Things to check
- [ ] Create Docker network via docker-compose

**Review by jakejack13 — COMMENTED** [review]
Docker port forwarding not working properly with Kafka and the workers. Will need to be fixed before merge

File: leaders/src/main/java/edu/cornell/worker/DockerWorker.java
Comment (owenxr) [reply leaders/src/main/java/edu/cornell/worker/DockerWorker.java:81]: Seems to work when local Kafka broker is set to 29092 instead of 9092

Comment (owenxr) [reply leaders/src/main/java/edu/cornell/worker/DockerWorker.java:81]: Fixed

**Comment by jakejack13** [comment]
Current problem: When workers are spawned and run inside of a Docker container, it cannot connect to Kafka when Kafka is run inside of a Docker container. Workers running on the host machine are able to connect to Kafka regardless of where Kafka is run. 

TODO
- [ ] Figure out why worker communication with host is blocked

Potential Solutions
- [ ] Set up a Docker network for all of the Docker containers to communicate with
- [ ] Change way Kafka is run (run on host?)

File: leaders/src/main/java/edu/cornell/worker/DockerWorker.java
Comment (owenxr) [inline leaders/src/main/java/edu/cornell/worker/DockerWorker.java:85]: currently, the kafkaAddress here is useless until structure is changed to use "kafka:29092" instead of the same thing the leader uses. If/When the leader is moved to using a docker container, the leader will need to use "kafka:29092" as well and be placed into the kafka_default network.

File: workers/src/main/java/edu/cornell/testoutputstream/KafkaTestOutputStream.java
Comment (owenxr) [inline workers/src/main/java/edu/cornell/testoutputstream/KafkaTestOutputStream.java:52]: currently hard coded kafkaAddress values

File: leaders/src/main/java/edu/cornell/worker/DockerWorker.java
Comment (jakejack13) [reply leaders/src/main/java/edu/cornell/worker/DockerWorker.java:85]: Should we move the `kafka` docker-compose file to top-level and add the leader to it? That way, everything boots up at the same time. We can configure the network on there

File: workers/src/main/java/edu/cornell/testoutputstream/KafkaTestOutputStream.java
Comment (jakejack13) [reply workers/src/main/java/edu/cornell/testoutputstream/KafkaTestOutputStream.java:52]: This should be added as a private constant to the top of the class (but it's fine for it to be constant for now)

File: workers/src/main/java/edu/cornell/testenv/testrunner/JUnit5TestRunner.java
Comment (jakejack13) [inline workers/src/main/java/edu/cornell/testenv/testrunner/JUnit5TestRunner.java:36]: Should this be included in the config so we don't have to hardcode it? Pretty sure we already have the list of test directories in the config

**Review by jakejack13 — COMMENTED** [review]
Small comment

File: workers/src/main/java/edu/cornell/testenv/testrunner/JUnit5TestRunner.java
Comment (owenxr) [reply workers/src/main/java/edu/cornell/testenv/testrunner/JUnit5TestRunner.java:36]: Ideally yes. Either in this PR or a separate small PR.

Comment (jakejack13) [reply workers/src/main/java/edu/cornell/testenv/testrunner/JUnit5TestRunner.java:36]: Let's do separate PR

File: leaders/src/main/java/edu/cornell/worker/DockerWorker.java
Comment (owenxr) [reply leaders/src/main/java/edu/cornell/worker/DockerWorker.java:85]: Let's experiment with this in a different PR because what we have works. Since all of these services depend on one another booting up our current system we know it works because we can wait for Kafka to be fully up then the leader etc. Although I suspect it should be fine to make this change, I think we should keep this working copy and then make that change separately.

File: workers/src/main/java/edu/cornell/testoutputstream/KafkaTestOutputStream.java
Comment (owenxr) [reply workers/src/main/java/edu/cornell/testoutputstream/KafkaTestOutputStream.java:52]: I will make this change.

**Comment by jakejack13** [comment]
LGTM for Owen's changes (waiting on review for my changes)

**Review by MitchellGray100 — APPROVED** [review]
LGTM

**Pull request merged**
